### ISSUE TITLE
docker is unable to start when /var/lib/docker/repositories-aufs is an empty file

### ISSUE THREAD

**Issue by stapelberg** [docker/docker#17411]
uname -a:
```
Linux gke-i3-faq-micro-e36b610a-node-4l9c 3.16.0-0.bpo.4-amd64 #1 SMP Debian 3.16.7-ckt11-1+deb8u2~bpo70+1 (2015-07-22) x86_64 GNU/Linux
```

docker info just spits the same error that docker version spits out at the end, likely because docker isn’t running:

docker version:
```
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64
FATA[0000] Get http:///var/run/docker.sock/v1.18/version: dial unix /var/run/docker.sock: no such file or directory. Are you trying to connect to a TLS-enabled daemon without TLS? 
```

Environment details:
Running on Google Container Engine

The disk of one of my VMs went full, but even after making some space, docker doesn’t start up again. For some reason, `/var/lib/docker/repositories-aufs` existed, but contained 0 bytes. I suspect docker directly writes to the file instead of creating a tempfile and renaming that tempfile — I think that should be fixed if it’s the case.

When starting docker using `service docker restart`, I see the following messages in `/var/log/docker.log`:
```
time="2015-10-27T19:25:03Z" level=info msg="+job serveapi(unix:///var/run/docker.sock)" 
time="2015-10-27T19:25:03Z" level=info msg="Listening for HTTP on unix (/var/run/docker.sock)" 
time="2015-10-27T19:25:03Z" level=fatal msg="Shutting down daemon due to errors: Couldn't create Tag store: unexpected end of JSON input" 
```

Once I delete `/var/lib/docker/repositories-aufs`, I get a different error message:
```
time="2015-10-27T19:24:49Z" level=info msg="+job serveapi(unix:///var/run/docker.sock)" 
time="2015-10-27T19:24:49Z" level=info msg="Listening for HTTP on unix (/var/run/docker.sock)" 
time="2015-10-27T19:24:49Z" level=fatal msg="Shutting down daemon due to errors: could not create trust store: error unmarshalling content: unexpected end of JSON input" 
```

Using `touch /var/lib/docker/repositories-aufs`, I can reproduce the issue and get the first error message again.

Can we make docker more robust in such a case and have it recreate the file on its own, or at least print an error message with clear instructions (e.g. “restore the file from backup, or delete if no backup”)?

**Comment by cpuguy83** [comment]
Ok, i can see that if the file is empty there is the json error.
But if the file does not exist, docker does create it.

**Comment by stapelberg** [comment]
Yes. What I’m trying to convey is that an empty file should be treated equal to “does not exist”.

**Comment by askb** [comment]
Hi, I was looking into a similar issue and would like to give this a go. If no one is working on this issue, #dibs :) 
Thanks

**Comment by askb** [comment]
When the file is deleted, the function `graph.store.reload()` tries to `os.Open()` conf file `/var/lib/docker/repositories-aufs` and returns error if file is missing. This scenario works fine as the `graph.store.save()` is invoked to re-create the file and daemon continues when the file is removed. 

However, in case where an empty file on-disk,  issue is caused whne `graph.store.reload()` returns error on trying to `Decode()` an empty file, instead of returning and  calling `graph.store.save()`, which stops the daemon from continuing.

Therefore, it would good to print a warning message to the logs in both cases and maybe continue loading the daemon for both cases.

**Comment by thaJeztah** [comment]
Looks like this issue is resolved by the new content addressability changes (https://github.com/docker/docker/pull/17924), see https://github.com/docker/docker/pull/17794#issuecomment-159572566

I'm going to close this issue, as this should be resolved in the upcoming 1.10 release, but please comment here if you're still able to reproduce on 1.10 once that has been released
